compiler/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -448,24 +433,25 @@ impl<'interner> TypeChecker<'interner> {
         bindings: TypeBindings,
         generics: Option<Vec<Type>>,
         function_generic_count: usize,
-        implicit_generic_count: usize,
         span: Span,
     ) -> (Type, TypeBindings) {
         match generics {
             Some(generics) => {
-                if generics.len() != (function_generic_count + implicit_generic_count) {
-                    // The list of `generics` represents the total number of generics on a function, including implicit generics.
-                    // We calculate any implicit generics earlier when type checking a method call,
-                    // In a user facing error we want to display only the number of function generics
-                    // as this is what the user will actually be specifying.
+                if generics.len() != function_generic_count {
                     self.errors.push(TypeCheckError::IncorrectTurbofishGenericCount {
                         expected_count: function_generic_count,
-                        actual_count: generics.len() - implicit_generic_count,
+                        actual_count: generics.len(),
                         span,
                     });
                     typ.instantiate_with_bindings(bindings, self.interner)
                 } else {
-                    typ.instantiate_with(generics)
+                    // Fetch the count of any implicit generics on the function, such as
+                    // for a method within a generic impl.
+                    let implicit_generic_count = match &typ {
+                        Type::Forall(generics, _) => generics.len() - function_generic_count,
+                        _ => 0,
+                    };
+                    typ.instantiate_with(generics, self.interner, implicit_generic_count)
```

Comment:
Since we're in the `else` case of `if generics.len() != function_generic_count`
the `generics.len() - function_generic_count` here should always be zero.

Reply:
Instead of checking `if generics.len() != function_generic_count` I think we need to check `if generics.len() != function_generic_count - implicit_generic_count`.

Reply:
`generics` should just contain the function generics as it is what is specified by the user in the Turbofish operator. `generics.len() != function_generic_count - implicit_generic_count` is what I was previously checking when I was adding the implicit generics to `generics`. I get the expected errors in `test_programs/execution_success/generics`:
<img width="712" alt="Screenshot 2024-05-21 at 9 41 31 AM" src="https://github.com/noir-lang/noir/assets/43554004/5be85b7d-0a64-4536-9324-41816a1d6e58">

Reply:
> generics.len() - function_generic_count here should always be zero.

`generics` has been re-scoped in the match case to the Forall generics which includes the implicit generics. This could be renamed if this is unclear though.

Reply:
> generics has been re-scoped in the match case to the Forall generics which includes the implicit generics

Thanks, I missed this. It could be helpful to rename the first generics to `turbofish_generics` maybe? I won't block on this though.